lib/dpdk/iceBoardVDIF.hpp

```diff
@@ -267,56 +359,87 @@ inline void iceBoardVDIF::handle_lost_samples(int64_t lost_samples) {
 
 void iceBoardVDIF::copy_packet_vdif(struct rte_mbuf* mbuf) {
 
-    const int64_t frame_size = vdif_packet_len * num_elements;
-
     int64_t vdif_frame_location = cur_seq - get_fpga_seq_num(out_buf, out_buf_frame_id);
 
-    if (unlikely((size_t)(vdif_frame_location * frame_size) == out_buf->frame_size)) {
+    if (unlikely((size_t)(vdif_frame_location * vdif_frameset_size) == out_buf->frame_size)) {
         advance_vdif_frame(cur_seq);
         vdif_frame_location = 0;
     }
 
-    assert((size_t)(vdif_frame_location * frame_size) <= out_buf->frame_size);
+    assert((size_t)(vdif_frame_location * vdif_frameset_size) <= out_buf->frame_size);
 
     // Set the VDIF headers only on the first port
     // since all ports would generate the same header
     if (port == 0) {
-        set_vdif_header_options(vdif_frame_location * frame_size, cur_seq);
+        set_vdif_header_options(vdif_frame_location * vdif_frameset_size, cur_seq);
     }
 
     auto& tel = Telescope::instance();
     stream_t encoded_id = ice_encode_stream_id(port_stream_id);
 
     // Create the parts of the VDIF frame that are in this packet.
-    int from_idx = header_offset + offset;
-    int mbuf_len = mbuf->data_len;
-    for (uint32_t time_step = 0; time_step < samples_per_packet; ++time_step) {
-        for (int freq = 0; freq < 128; ++freq) {
-            for (int elem = 0; elem < num_elements; ++elem) {
-
-                // Advance to the next mbuf in the chain.
-                if (unlikely(from_idx >= mbuf_len)) {
-                    mbuf = mbuf->next;
-                    assert(mbuf);
-                    from_idx -= mbuf_len; // Subtract the last mbuf_len from the current idx.
-                    mbuf_len = mbuf->data_len;
+    char* mbuf_data_base = rte_pktmbuf_mtod(mbuf, char*);
+    char* mbuf_data_max = mbuf_data_base + mbuf->data_len;
+    char* mbuf_data_ptr = mbuf_data_base + header_offset;
+    uint8_t* out_t0_f0 =
+        (out_buf_frame                              // Start of output buffer.
+         + vdif_frame_location * vdif_frameset_size // Frame location in output buffer.
+         + vdif_header_len);                        // Offset for the vdif header.
+
+    // Offset in output frame for first frequency of this receiver buffer thread (0-7).
+    uint32_t freq_offset = tel.to_freq_id(encoded_id, 0);
+
+    // Times are in separate frames, so time stride equals the size of the VDIF framesets.
+    for (uint8_t* out_t_f0 = out_t0_f0;
+         out_t_f0 < out_t0_f0 + vdif_frameset_size * samples_per_packet;
+         out_t_f0 += vdif_frameset_size) {
+        // Pointer to start of thread in VDIF frame (updated in loop).
+        uint8_t* out_t_p = out_t_f0;
+        for (uint32_t i_thread = 0; i_thread < num_threads; i_thread++) {
+            // Pointer to starting input and frequency in input buffer for this thread.
+            char* in = mbuf_data_ptr + buffer_offsets[i_thread];
```

Comment:
I'm worried that `in` might exceed `mbuf_data_max`.   Have you checked in all cases below that the `in` pointer never reads past the end of the current `mbuf` (for each of the sizes it can be cast to)?   It may actually be possible that it just happens to work out because of how our packet size, the mbuf sizes, and headers line up, but generically I'm not sure this should work, and would like to double check.

Reply:
Yikes, yes, it does, and some debug statements show the buffer size is 2048, of which 58 is header, so 1990 left, which is indeed not even a multiple of 16... I think I have an idea how to do it, using that it won't be split over more than 2 frames, but clearly this is not as simple as I thought...

Reply:
OK, now completely rewrote. As I realized I could not count on things being properly aligned, I tested just using `std::copy_n` and it goes just as fast without trying to be clever and have fixed `num_elements`. So, the whole is much simpler now, even with the partial samples.